python/taichi/lang/__init__.py

```diff
@@ -71,22 +71,55 @@ def reset():
     runtime = get_runtime()
 
 
+class _EnvironConfig:
+    def __init__(self, args):
+        self.args = args
+        self.cfg_keys = []
+
+    def add(self, key, default=None, cast=None):
+        cast = cast or self.bool_int
+
+        self.cfg_keys.append(key)
+        # TI_ASYNC=   : no effect
+        # TI_ASYNC=0  : False
+        # TI_ASYNC=1  : True
+        name = 'TI_' + key.upper()
+        value = os.environ.get(name, '')
+        if len(value):
+            self.args[key] = cast(value)
+        elif key not in self.args:
+            self.args[key] = default
```

Comment:
Sorry this is not what I meant. Doing this will still cause the default value to be defined in two places. Instead, what I suggested is to remove `default` completely and do this:

```suggestion
            self.args[key] = getattr(ti.cfg, key)
```

This way, we know for sure the default value comes from `CompileConfig`.

Reply:
`NameError: ti.cfg.log_level not found`.

Reply:
Ah, but i guess there could be the case where `key` is not part of `cfg`, so that `default` is still needed? Then maybe

```py
try:
  self.args[key] = getattr(ti.cfg, key)
except:
  self.args[key] = default  # |key| should be things like `print_preprocessed`, `gdb_trigger`, etc.
```

Reply:
This sounds too ad-hoc to me. Please see https://github.com/taichi-dev/taichi/pull/1396#discussion_r450269970.